### PR TITLE
[WIP] Module Slicing 2

### PR DESCRIPTION
**Pull request by alexander-g**
Alternative approach to #115 
- This is now based on `jax.make_jaxpr` and `jax.named_call`, not on summaries. Specifically, I've extended hooks context with a new attribute, if it is active, `Module.call` will be wrapped with `jax.named_call`. This allows identifying individual modules in the low-level jaxpr commands, so the interface is basically the same as before. However, one can now use arbitrary JAX functions without having to wrap them in modules:
```python
class Module(elegy.Module):
    def call(self, x):
        x = x/255.
        x = elegy.nn.Linear(300, name="linear0")(x)
        x = jax.nn.relu(x)
        x = elegy.nn.Linear(100, name="linear1")(x)
        x = jax.nn.relu(x)
        x = elegy.nn.Linear(10,  name="linear2")(x)
        return x

x         = np.random.random([32,1024])
model     = elegy.Model(Module())
submodule = elegy.slicing.slice_model(model, start='input', end=['linear0', 'linear1'], sample_input=x)
out       = elegy.Model(submodule).predict(x)

assert out[0].shape == (32,300)
assert out[1].shape == (32,100)
```
- Some limitations still apply:
  - Only Elegy Modules are possible as start and end targets
  - Only one start target allowed. This is possible to fix, but might get complex
- Internal logic is simpler, not graph-based as before
- WIP, still need to cover some edge cases and test more

### CODE REVIEW

File: elegy/slicing.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+
+def slice_model(
+    model: elegy.Model,
+    start: tp.Union[elegy.Module, str, None],
+    end: tp.Union[elegy.Module, str, None, tp.List[tp.Union[elegy.Module, str, None]]],
+    sample_input: np.ndarray,
+) -> elegy.Model:
+    
+    model.maybe_initialize(elegy.types.Mode.pred, x=sample_input)
```
Comment (cgarciae) [inline elegy/slicing.py:20]: When you update to master maybe you can check for `model.initialized` since `maybe_initialize` is gone.

**Comment by cgarciae** [comment]
Thanks @alexander-g!

I have one main question: are the parameters being transferred to the new module thanks to `Module.set_default_parameters`?

I am assuming the answer is yes which is fine, however, if this is so, should this method be part of `elegy.Module` instead? This might make more sense now that we support other module systems.

Due to recent changes your example might require some minor modifications:
* To apply `set_default_parameters` to an elegy `Module` from `Model` use `model.update_module()`
* To initialize the `Model` on predict you can use `mode.predict(x, initialize=True)`

**Comment by alexander-g** [comment]
> I have one main question: are the parameters being transferred to the new module thanks to Module.set_default_parameters?

Parameters are transferred but I do not call `set_default_parameters` explicitly. I construct a new Module with the old modules as submodules. Not sure how exactly this works but it does.

> should this method be part of elegy.Module instead?

Yes, I've added `Model.slice()`


> To initialize the new Model on predict you can use model.predict(x, initialize=True)

This is a bit annoying and counterintuitive that I have to initialize a pretrained Model. Can this be avoided?

***
Ready for review.
How to I trigger the tests?

**Comment by cgarciae** [comment]
> Parameters are transferred but I do not call set_default_parameters explicitly. I construct a new Module with the old modules as submodules. Not sure how exactly this works but it does.

Interesting. In the latest version I think you have to be more explicit about this but its just a single line of code.

> This is a bit annoying and counterintuitive that I have to initialize a pretrained Model. Can this be avoided?

Ok so there is a pathological case here I'll show next so I coded a "defensive" solution but I am open to suggestions:

We conditionally initialize on `predict`, `evaluate`, `fit` based on whether Model is initialize or not and using on the available information for that method (x, y, sample_weights, etc) so if you have losses and metrics doing this is ok:

```python
model.fit(X_train, y_train)
preds = model.predict(X_train)
```
but doing this will raise exceptions on initialization since the losses and metrics don't have the labels:
```python
preds = model.predict(X_train)
model.fit(X_train, y_train)
```
So the proposed solution is to have the user opt-in via `initialize=True` and raise a friendly error if he didn't with information on how it can fixed the issue instead of just letting the program crash. All this stems from #165.

**Comment by cgarciae** [comment]
@alexander-g Now that its not a draft I think just commit a random change.

**Comment by alexander-g** [comment]
> Interesting. In the latest version I think you have to be more explicit about this but its just a single line of code.

You're right, it's required, I've added it. It had worked before simply because of the same random seed.

**Comment by cgarciae** [comment]
@alexander-g is this ready for review? :)

**Comment by alexander-g** [comment]
Yes it is.
Just added some comments in the inner logic.

**Comment by cgarciae** [comment]
Hey @alexander-g! Sorry for the hiatus, a couple of change important changes recently kept me busy, good new is that my new employer is sponsoring a couple of hours a week for me to work on Elegy continuously :)

Regarding this PR, I definitely thinks its best that this is a feature of `elegy.Module` instead of `elegy.Model` since this won't work with any other Module system.

**Comment by alexander-g** [comment]
- moved from `Model` to `Module`. The user has to call `model.update_modules()` manually to ensure parameters are transferred.
